Explain event callbacks in "Monitoring for parameter changes" (#5543)
ros2/ros2_documentation: source/Tutorials/Intermediate/Monitoring-For-Parameter-Changes-CPP.rst

The callback we set previously in the node has been invoked and has displayed the new updated value.
You can now terminate the running parameter_event_handler sample using ^C in the terminal.

> 

Reviewer: Nitpick: only 1 blank line between sections/anything, like before.

```suggestion
```

Extensions
----------

So far, we built and tested a small node that monitors a single parameter owned by the node itself.

> Using this node as a base, two other usecases where the ParameterEventHandler can be useful is presented below.

Reviewer: ```suggestion
Using this node as a base, two other usecases where the ParameterEventHandler can be useful are presented below.
```

> 

Reviewer: Nitpick: keep 1 blank line between sections, like before.

```suggestion
```

    [INFO] [1606952588.237531933] [node_with_parameters]: cb2: Received an update to parameter "a_double_param" of type: double: "3.45"

> 

Reviewer: Nitpick: keep 1 blank line between sections, like before.

```suggestion
```

Monitor all node parameters simultaneously
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

If you need to monitor multiple nodes or parameters at the same time, it would be cumbersome to have to call ``add_parameter_callback`` once for each of them.

> In this case, you can use ``add_parameter_event_callback`` to register a single callback that fires when *any* of the node's parameters change.

Reviewer: Same question here: is this only for a specific node's parameters, or is it for all parameters of all nodes?

To do this, first update the SampleNodeWithParameters constructor to add the following code:

.. code-block:: C++

    this->declare_parameter("another_double_param", 0.0);

    ...

    auto event_cb = [this](const rcl_interfaces::msg::ParameterEvent & parameter_event) {
        RCLCPP_INFO(this->get_logger(), "Received parameter event");

        for (const auto& p : parameter_event.changed_parameters) {
          RCLCPP_INFO(
            this->get_logger(), "Inside event: \"%s\" changed to %s",
            p.name.c_str(),
            rclcpp::Parameter::from_parameter_msg(p).value_to_string().c_str());
        };
      };
    event_cb_handle_ = param_subscriber_->add_parameter_event_callback(event_cb);

> 

Reviewer: ```suggestion
```

This declares a new double parameter ``another_double_param`` and adds an event callback that will monitor both parameters.

> Note that the ``parameter_event`` is of type `rcl_interfaces/ParameterEvent <https://docs.ros.org/en/rolling/p/rcl_interfaces/msg/ParameterEvent.html>`_.

Reviewer: ```suggestion
Note that the ``parameter_event`` is of type {interface(rcl_interfaces/msg/ParameterEvent)}.
```

Although it's not shown in this tutorial, event callbacks can also be used to monitor when parameters are added or deleted.

Finally, don't forget to add the event callback handle as a private member:

.. code-block:: C++

    private:
      ...
      std::shared_ptr<rclcpp::ParameterEventCallbackHandle> event_cb_handle_;

> 

Reviewer: ```suggestion
```

Navigate back to the root of your workspace, ``ros2_ws``, and rebuild your updated package as before:

.. code-block:: console

    $ colcon build --packages-select cpp_parameter_event_handler

Then source the setup files:

.. tabs::

  .. group-tab:: Linux

    .. code-block:: console

      $ . install/setup.bash

  .. group-tab:: macOS

    .. code-block:: console

      $ . install/setup.bash

  .. group-tab:: Windows

    .. code-block:: console

      $ call install\setup.bat

> 

Reviewer: ```suggestion
```

To test the new event callback, first run the parameter_event_handler node:

.. code-block:: console

     $ ros2 run cpp_parameter_event_handler parameter_event_handler

> Then, from a second terminal (with ROS initialized), let's set the original int parameter:

Reviewer: I'm assuming you mean "sourcing," so it would be better to say that, since that's what we call it everywhere:

```suggestion
Then, from a second terminal (with ROS sourced), let's set the original int parameter:
```

.. code-block:: console

     $ ros2 param set node_with_parameters an_int_param 44

Upon executing this command, you should see both the single-parameter callback, as well as the event callback being fired:

.. code-block:: console

      [INFO] [1747144403.418980063] [node_with_parameters]: cb: Received an update to parameter "an_int_param" of type integer: "44"
      [INFO] [1747144403.419086611] [node_with_parameters]: Received parameter event
      [INFO] [1747144403.419114103] [node_with_parameters]: Inside event: "an_int_param" changed to 44

> 

Reviewer: ```suggestion
```

Now set the new double parameter:

.. code-block:: console

     $ ros2 param set node_with_parameters another_double_param 4.4

Since no single-parameter callback was added (via ``add_parameter_callback``) for the double parameter, we should see only the event callback fire:

.. code-block:: console

      [INFO] [1747144452.917437113] [node_with_parameters]: Received parameter event
      [INFO] [1747144452.917591649] [node_with_parameters]: Inside event: "another_double_param" changed to 4.400000

.. note::

   When setting multiple parameters at once, it's best to use ``set_parameters_atomically``, explained in :doc:`../../Concepts/Basic/About-Parameters`.
   This way, the event callback is only fired once.

> 

Reviewer: ```suggestion
```
